### PR TITLE
Add Shell badge support (BadgeText, BadgeColor, BadgeTextColor)

### LINKED ISSUE #8305: Shell badges
### Description

any idea if this is going to be implemented?
https://github.com/xamarin/XamarinCommunityToolkit/issues/517

### Public API Changes

https://github.com/xamarin/XamarinCommunityToolkit/issues/517

### Intended Use-Case

that Shell has by default add badges, both in TabBar and in the floating menu.

### PR DESCRIPTION
**Pull request by jfversluis**
> [!NOTE]
> Are you waiting for the changes in this PR to be merged?
> It would be very helpful if you could <a href="https://github.com/dotnet/maui/wiki/Testing-PR-Builds">test the resulting artifacts</a> from this PR and let us know in a comment if this change resolves your issue. Thank you!

## Description

Adds badge notification support to Shell bottom tabs on all platforms (Android, iOS, MacCatalyst, Windows).

Closes #8305 (first iteration — bottom tabs)

### New API

Three new bindable properties on `BaseShellItem`:

| Property | Type | Description |
|----------|------|-------------|
| `BadgeText` | `string?` | Text/number displayed on the badge. `null` = no badge, `""` = dot indicator, non-empty = text/count badge. |
| `BadgeColor` | `Color?` | Background color of the badge. `null` uses the platform default. |
| `BadgeTextColor` | `Color?` | Foreground (text) color of the badge. `null` uses the platform default (typically white). |

### XAML Usage

```xml
<Shell>
    <TabBar>
        <ShellContent Title="Home" BadgeText="3" BadgeColor="Red" BadgeTextColor="White" />
        <ShellContent Title="Alerts" BadgeText="" /> <!-- dot indicator -->
    </TabBar>
</Shell>
```

### Platform Implementation

| Platform | Approach | Notes |
|----------|----------|-------|
| **Android** | `BottomNavigationView.GetOrCreateBadge()` / `BadgeDrawable` | Full text + color support. `ClearNumber()` for dot mode. |
| **iOS/MacCatalyst** | `UITabBarItem.BadgeValue` / `BadgeColor` / `SetBadgeTextAttributes` | Full text + color support. Empty string shows native dot. |
| **Windows** | `InfoBadge` on `NavigationViewItem` via ViewModel binding | Numeric values show number; non-numeric/empty shows dot indicator. |

### Dot Badge Support

Setting `BadgeText = ""` (empty string) displays a small dot indicator without text on all platforms. This is distinct from `null` which hides the badge entirely.

### What's Included

- 3 BindableProperties on `BaseShellItem` with full XML documentation
- Platform handlers for Android, iOS/MacCatalyst, and Windows
- 28 unit tests (`ShellBadgeTests.cs`)
- 4 UI screenshot tests (`Issue8305.cs`)
- Interactive sample gallery page (`ShellBadgeGallery`)
- PublicAPI.Unshipped.txt entries for all 7 TFMs

### What's NOT Included (follow-up work)

- Flyout item badges
- Top tab badges
- These are tracked in the original issue #8305 for future iterations

### Screenshots

screen12

### Testing

- All 28 unit tests pass
- All 85 existing Shell unit tests pass (no regression)
- UI test projects compile successfully
- Manually tested on physical Android device

### CODE REVIEW

File: src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellItemRenderer.cs [hunk 1]
```diff
@@ -497,6 +510,56 @@ protected virtual void UpdateShellSectionEnabled(ShellSection shellSection, IMen
...
+				return;
+
+			var items = ((IShellItemController)ShellItem).GetItems();
+			var maxItems = _bottomView.MaxItemCount;
+			var count = Math.Min(items.Count, maxItems);
+
+			for (int i = 0; i < count; i++)
+			{
+				UpdateShellSectionBadge(items[i], i);
+			}
```
[the remark this answers is not in the document]
Comment (jfversluis) [reply src/Controls/src/Core/Compatibility/Handlers/Shell/Android/ShellItemRenderer.cs]: Fixed in 23b14ca. Updated `UpdateAllBadges()` to skip the More tab index when overflow exists (`lastIndexToUpdate = maxItems - 2`), and explicitly clear any badge on the More tab via `RemoveBadge(maxItems - 1)`.

File: src/Controls/src/Core/Shell/BaseShellItem.cs [hunk 2]
```diff
@@ -149,6 +157,29 @@ public bool FlyoutItemIsVisible
...
+		}
+
+		/// <summary>
+		/// Gets or sets the background color of the badge displayed on this Shell navigation item. This is a bindable property.
+		/// </summary>
+		public Color BadgeColor
+		{
+			get => (Color)GetValue(BadgeColorProperty);
+			set => SetValue(BadgeColorProperty, value);
+		}
```
[the remark this answers is not in the document]
Comment (jfversluis) [reply src/Controls/src/Core/Shell/BaseShellItem.cs:186]: This file has `#nullable disable` at line 1, which is intentional and consistent with all other properties in `BaseShellItem` (e.g. `Title`, `Route`, `Icon` — all follow the same oblivious pattern). In nullable-enabled consumers, assigning `null` to an oblivious-typed property does not produce CS8625 — oblivious types suppress nullable warnings by design. The PublicAPI entries correctly use the `~` prefix to denote oblivious nullability. No change needed here.

File: src/Controls/samples/Controls.Sample/Pages/Core/ShellGalleries/ShellBadgeGallery.cs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+		}
+
+		IList<ShellSection> GetAllShellSections()
+		{
+			var shellItem = Shell.Current?.CurrentItem;
+			if (shellItem is null)
+				return null;
+
+			return ((IShellItemController)shellItem).GetItems();
+		}
```
[the remark this answers is not in the document]
Comment (jfversluis) [reply src/Controls/samples/Controls.Sample/Pages/Core/ShellGalleries/ShellBadgeGallery.cs:227]: Fixed in 23b14ca. Updated nullable annotations: `Color?` for `CreateColorButton`, `ShellSection?` for `GetCurrentShellSection()`, and `IList<ShellSection>?` for `GetAllShellSections()`.

File: src/Controls/src/Core/PublicAPI/net-tizen/PublicAPI.Unshipped.txt [hunk 4]
```diff
@@ -24,6 +28,8 @@ Microsoft.Maui.Controls.LongPressingEventArgs
 Microsoft.Maui.Controls.LongPressingEventArgs.LongPressingEventArgs(Microsoft.Maui.GestureStatus status) -> void
 Microsoft.Maui.Controls.LongPressingEventArgs.Status.get -> Microsoft.Maui.GestureStatus
 override Microsoft.Maui.Controls.Shapes.Shape.OnPropertyChanged(string? propertyName = null) -> void
+static readonly Microsoft.Maui.Controls.BaseShellItem.BadgeColorProperty -> Microsoft.Maui.Controls.BindableProperty!
+static readonly Microsoft.Maui.Controls.BaseShellItem.BadgeTextProperty -> Microsoft.Maui.Controls.BindableProperty!
```
[the remark this answers is not in the document]
Comment (jfversluis) [reply src/Controls/src/Core/PublicAPI/net-tizen/PublicAPI.Unshipped.txt]: Fixed in 23b14ca. Aligned net-tizen entries with all other TFMs — now uses `~` prefix for oblivious nullability (matching the `#nullable disable` in `BaseShellItem.cs`) and removed `?`/`!` annotations.

**Comment by AlleSchonWeg** [comment]
@jfversluis ,
thanks for this addition. Is it possible to also add this on a `Microsoft.Maui.Controls.ToolbarItem`? I know that is supported on iOS 26 (UIBarButtonItemBadge) and Android (BadgeDrawable).

**Comment by kubaflo** [comment]
## Code Review Summary

**Shell badge tabs support** — Adds `BadgeText`, `BadgeColor`, `BadgeTextColor` to `BaseShellItem` with platform-specific rendering on Android, iOS/MacCatalyst, and Windows. 28 unit tests, 4 UI screenshot tests, comprehensive gallery sample.

### Findings

| Severity | Finding |
|----------|---------|
| 🔴 Critical | **Android: Missing `index < 0` guard** in `OnShellSectionPropertyChanged` — `IndexOf` returns -1 if a `ShellSection` is removed while `PropertyChanged` fires → `GetOrCreateBadge(-1)` or `RemoveBadge(-1)` will throw. Fix: add `if (index < 0) return;` before the overflow guard |
| 🟡 Warning | Android `UpdateShellSectionBadge`: confusing "remove then recreate" pattern when `badgeColor` is null — works but intent is unclear |
| 🟡 Warning | iOS uses private static `UpdateTabBarItemBadge` vs Android's `protected virtual` — extensibility inconsistency for custom Shell renderers |
| 🟡 Warning | Windows non-numeric badge text silently degrades to dot — documented but no runtime diagnostic |
| 🟡 Warning | Tizen `PublicAPI.Unshipped.txt` missing trailing newline |
| 🔵 Info | PublicAPI diffs include unrelated re-sorting (inflates diff) |
| 🔵 Info | Excellent test coverage (28 unit + 4 UI screenshot tests) |
| 🔵 Info | Clean API design extending `BaseShellItem` naturally |

### Overall Assessment

✅ **Looks good** — well-designed feature with excellent test coverage. The Android index guard (🔴) should be fixed to prevent crashes on section removal during property change propagation. Other findings are non-blocking.

---
*Review performed by Copilot CLI*

**Comment by kubaflo** [comment]
## Code Review — PR #34659

### Independent Assessment
**What this changes:** Adds `BadgeText`, `BadgeColor`, and `BadgeTextColor` bindable properties to `BaseShellItem`, enabling badge notifications on Shell bottom tabs across Android, iOS/MacCatalyst, and Windows. Companion to PR #34669 (ToolbarItem badges) — together they close #8305.

**Target branch:** `net11.0` — correct for new API feature. Milestone `.NET 11.0-preview4`.

### Reconciliation with PR Narrative
**Agreement:** ✅ Code matches description. Three new properties on `BaseShellItem` with platform implementations using native badge APIs. Same null/empty/text semantics as the ToolbarItem PR.

---

### Findings

#### ✅ Good — Consistent API design with ToolbarItem badges
Same three properties (`BadgeText`, `BadgeColor`, `BadgeTextColor`) with identical semantics:
- `null` = hidden, `""` = dot indicator, non-empty = text/count
- Properties use `BindingMode.OneWay` — appropriate for display-only properties set from code/binding
- Comprehensive XML documentation on all three properties

#### ✅ Good — Android: Correct overflow handling
`UpdateAllBadges()` correctly handles the "More" tab overflow case:
- When `items.Count > maxItems`, indices `0..maxItems-2` get badges, index `maxItems-1` (More tab) is explicitly cleared
- `OnShellSectionPropertyChanged` has the same guard: `!(itemCount > maxItems && index > maxItems - 2)`
- `UpdateShellSectionBadge` uses `BottomNavigationView.GetOrCreateBadge()`/`RemoveBadge()` — native Material Design API

#### ✅ Good — iOS: Clean native integration
```csharp
tabBarItem.BadgeValue = badgeText is null ? null : (badgeText.Length > 0 ? badgeText : "");
```
Single-line mapping to `UITabBarItem.BadgeValue` with correct semantics. `BadgeColor` and `SetBadgeTextAttributes` for color customization. No version check needed — `UITabBarItem.BadgeValue` has been available since iOS 2.0. Static helper method keeps code clean.

#### ✅ Good — Windows: Proper MVVM approach
Badge support via ViewModel binding rather than imperative code:
- `NavigationViewItemViewModel` extended with `BadgeText`, `BadgeBackground`, `BadgeForeground`, computed `BadgeValue`, `BadgeVisibility`, `HasBadge`
- XAML template binds `InfoBadge.Value`, `Background`, `Foreground`, `Visibility` declaratively
- `_badgeTrackedSections` HashSet manages `PropertyChanged` subscriptions with cleanup in `DisconnectHandler` and `SetVirtualView`
- Smart `BadgeValue` computation: `int.TryParse` → numeric count, else `-1` → WinUI dot indicator

#### ✅ Good — Test coverage
- 28 unit tests covering defaults, get/set, clearing, PropertyChanged, binding, edge cases
- 4 UI screenshot tests: initial badge, runtime update, clearing, multiple tabs
- Screenshots for Android, iOS 26, and Windows
- Gallery sample page (`ShellBadgeGallery`) with interactive controls

#### ✅ Good — All prior review feedback addressed
All 4 review threads resolved:
| Concern | Resolution |
|---------|-----------|
| More tab overflow badge | Fixed — skips maxItems-1, explicitly clears More badge |
| Nullable annotations in BaseShellItem | `#nullable disable` is intentional, matches existing properties |
| Nullable annotations in sample gallery | Fixed — `Color?`, `ShellSection?`, `IList<ShellSection>?` |
| Tizen PublicAPI alignment | Fixed — uses `~` prefix consistently |

---

#### 💡 Observation — Android remove+recreate for default color reset
In `UpdateShellSectionBadge`, when `badgeText` is non-null but `badgeColor` is null:
```csharp
// Remove and recreate badge when clearing color to reset to platform default
if (badgeColor is null)
{
    _bottomView.RemoveBadge(menuItemId);
}
var badge = _bottomView.GetOrCreateBadge(menuItemId);
```
This means every badge without a custom color does a remove+recreate cycle. The comment explains the intent (Material `BadgeDrawable` has no "reset to default" API). This is correct but creates a brief flash on initial badge creation when no custom color is set. If this becomes a visual issue, the `badgeColor is null` removal could be skipped when the badge doesn't already exist. Not a blocker.

#### 💡 Observation — Windows `BadgeValue` semantics
`BadgeValue` returns `-1` for non-numeric/empty text, which WinUI `InfoBadge` renders as a dot. For negative numeric strings (e.g., `"-5"`), `int.TryParse` succeeds but the `v >= 0` guard catches it, so it shows as a dot. This is correct — negative badge counts don't make sense.

#### ℹ️ Note — Flyout and top tab badges are out of scope
The PR explicitly notes these are tracked for future iterations in #8305. Bottom tabs only for this iteration.

#### ℹ️ Note — CI Status
No CI failures detected. Clean build.

---

### Devil's Advocate
- *"Should badge be on `BaseShellItem` or `ShellSection`?"* — `BaseShellItem` is correct. It's the base for all navigable shell items, and future iterations (flyout badges, top tabs) will need it at this level.
- *"Why not use a `BadgeMode` enum instead of string semantics?"* — The string approach matches iOS native (`UITabBarItem.BadgeValue` is a string) and is more flexible (supports "99+", "New", etc.). An enum would be too restrictive.
- *"Could the BottomNavigationView badge flicker on remove+recreate?"* — Theoretically, but `RemoveBadge` + `GetOrCreateBadge` happens synchronously on the UI thread, so the flicker should be imperceptible. If reports arise, a check for existing badge before removal could be added.
- *"Why doesn't iOS need a version check like the ToolbarItem PR?"* — `UITabBarItem.BadgeValue` has been available since iOS 2.0. The ToolbarItem PR uses `UIBarButtonItem.badge` which is iOS 26+ only.

### Verdict: LGTM
**Confidence:** high  
**Summary:** Well-designed feature with consistent API across Shell and ToolbarItem badge systems. All three platform implementations use the correct native APIs. The Android overflow handling, iOS one-liner mapping, and Windows MVVM binding approach are all clean. 28 unit tests + 4 UI tests provide solid coverage. All prior review feedback addressed. Ready for merge alongside the companion PR #34669.

*Review performed by Copilot CLI using the code-review skill*

**Review by kubaflo — APPROVED** [review]
